ArrayList/DoubleLinkedList.cs

```diff
@@ -4,7 +4,752 @@
+        }
+
+        public void AddByIndex(T value, int index)
+        {
+            if (index >= 0 && index < Length)
+            {
+                if (index == 0)
+                {
+                    AddFirst(value);
+                }
+                else
+                {
+                    DoubleLinkedListNode<T> insertNode = new DoubleLinkedListNode<T>(value);
+                    DoubleLinkedListNode<T> currentNode = GetNodeByIndex(index - 1);
+
+                    insertNode.Next = currentNode.Next;
+                    currentNode.Next = insertNode;
+                    insertNode.Previous = currentNode;
+                    insertNode.Next.Previous = insertNode;
+                    ++Length;
+                }
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        public void AddList(DoubleLinkedList<T> list)
+        {
+            if (list.Length != 0)
+            {
+                if (_tail is null)
+                {
+                    _root = list._root;
+                }
+                else
+                {
+                    list._root.Previous = _tail;
+                    _tail.Next = list._root;
+                    _tail = list._tail;
+                }
+
+                Length += list.Length;
+            }
+        }
+
+        public void AddListFirst(DoubleLinkedList<T> list)
+        {
+            if(Length == 0)
+            {
+                _root = list._root;
+                _tail = list._tail;
+                Length = list.Length;
+            }
+            else if (list.Length != 0)
+            {
+                _root.Previous = list._tail;
+                list._tail.Next = _root;
+                _root = list._root;
+                Length += list.Length;
+            }
+        }
+
+        public void AddListByIndex(DoubleLinkedList<T> list, int index)
+        {
+            if (index >= 0 && index < Length)
+            {
+                if (index == 0)
+                {
+                    AddListFirst(list);
+                }
+                else
+                {
+                    DoubleLinkedListNode<T> currentNode = GetNodeByIndex(index - 1);
+                    list._root.Previous = currentNode;
+                    currentNode.Next.Previous = list._tail;
+                    list._tail.Next = currentNode.Next;
+                    currentNode.Next = list._root;
+                    Length += list.Length;
+                }
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
+            }
+        }
+        
+        public DoubleLinkedList<T> Sort(bool isDescending)
+        {
+            DoubleLinkedList<T> list = new DoubleLinkedList<T>();
+            list._root = this._root;
+            list._tail = this._tail;
+            list.Length = this.Length;
+            return MergeSort(list, isDescending);
```

Comment:
You can calc `coef` here and then send it to private methods

Reply:
Corrected